File: app/src/main/java/com/king/mvvmframe/app/adapter/BindingAdapter.java
package com.king.mvvmframe.app.adapter;

import android.content.Context;


import com.king.base.adapter.BaseRecyclerAdapter;
import com.king.mvvmframe.BR;

import java.util.List;

import androidx.annotation.LayoutRes;
import androidx.databinding.ViewDataBinding;

/**
 * @author <a href="mailto:jenly1314@gmail.com">Jenly</a>
 */
public class BindingAdapter<T> extends BaseRecyclerAdapter<T,BindingHolder<ViewDataBinding>> {

    public BindingAdapter(Context context, int layoutId) {
        super(context, layoutId);
    }

    public BindingAdapter(Context context, List<T> listData, @LayoutRes int layoutId) {
        super(context, listData,layoutId);
    }

    @Override
    public void bindViewDatas(BindingHolder<ViewDataBinding> holder, T item, int position) {
        if(holder.mBinding != null){
            holder.mBinding.setVariable(BR.data,item);
            holder.mBinding.executePendingBindings();
        }
    }

    public T getItem(int position) {
        if(position<getItemCount()){
            return getListData().get(position);
        }

        return null;
    }

    public void refreshData(List<T> list){
        if(list!=null){
            setListData(list);
        }else{
            getListData().clear();
        }
        notifyDataSetChanged();

    }
}

File: app/src/main/java/com/king/mvvmframe/app/adapter/BindingAdapter.kt
package com.king.mvvmframe.app.adapter

import android.annotation.SuppressLint
import android.content.Context
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.DiffUtil
import com.king.base.adapter.BaseRecyclerAdapter
import com.king.mvvmframe.BR

/**
 * @author <a href="mailto:jenly1314@gmail.com">Jenly</a>
 * <p>
 * <a href="https://github.com/jenly1314">Follow me</a>
 */
class BindingAdapter<T> : BaseRecyclerAdapter<T, BindingHolder<ViewDataBinding>> {
    constructor(context: Context, layoutId: Int) : super(context, layoutId)
    constructor(context: Context, listData: MutableList<T>?, layoutId: Int) : super(
        context,
        listData,
        layoutId
    )

    override fun bindViewDatas(holder: BindingHolder<ViewDataBinding>, item: T, position: Int) {
        holder.binding?.also {
            it.setVariable(BR.data, item)
            it.executePendingBindings()
        }
    }

    fun getItem(position: Int): T? {
        return listData.getOrNull(position)
    }

    fun refreshData(list: List<T>?) {
        val diffResult = DiffUtil.calculateDiff(ItemDiffCallback(listData, list.orEmpty()))
        if (list != null) {
            setListData(list)
        } else {
            listData.clear()
        }
        diffResult.dispatchUpdatesTo(this)
    }

    class ItemDiffCallback<T>(
        private val oldList: List<T>,
        private val newList: List<T>,
    ) : DiffUtil.Callback() {
        override fun getOldListSize(): Int {
            return oldList.size
        }

        override fun getNewListSize(): Int {
            return newList.size
        }

        override fun areItemsTheSame(oldPosition: Int, newPosition: Int): Boolean {
            return oldList[oldPosition] == newList[newPosition]
        }

        override fun areContentsTheSame(oldPosition: Int, newPosition: Int): Boolean {
            return oldList[oldPosition] == newList[newPosition]
        }
    }
}
